crates/ide_assists/src/handlers/inline_call.rs

```diff
@@ -95,41 +102,94 @@ pub(crate) fn inline_(
         label,
         expr.syntax().text_range(),
         |builder| {
-            // FIXME: emit type ascriptions when a coercion happens?
-            // FIXME: dont create locals when its not required
-            let statements = new_bindings
-                .map(|(pattern, value)| make::let_stmt(pattern, Some(value)).into())
-                .chain(body.statements());
+            let body = body.clone_for_update();
+
+            let file_id = file_id.original_file(ctx.sema.db);
+            let usages_for_locals = |local| {
+                Definition::Local(local)
+                    .usages(&ctx.sema)
+                    .all()
+                    .references
+                    .remove(&file_id)
+                    .unwrap_or_default()
+                    .into_iter()
+            };
+            // Contains the nodes of usages of parameters.
+            // If the inner Vec for a parameter is empty it either means there are no usages or that the parameter
+            // has a pattern that does not allow inlining
+            let param_use_nodes: Vec<Vec<_>> = params
+                .iter()
+                .map(|(pat, param)| {
+                    if !matches!(pat, ast::Pat::IdentPat(pat) if pat.is_simple_ident()) {
+                        return Vec::new();
+                    }
+                    usages_for_locals(param.as_local(ctx.sema.db))
+                        .map(|FileReference { name, range, .. }| match name {
+                            ast::NameLike::NameRef(_) => body
+                                .syntax()
+                                .covering_element(range)
+                                .ancestors()
+                                .nth(3)
+                                .filter(|it| ast::PathExpr::can_cast(it.kind())),
+                            _ => None,
+                        })
+                        .collect::<Option<Vec<_>>>()
+                        .unwrap_or_default()
+                })
+                .collect();
+
+            // Rewrite `self` to `this`
+            if param_list.self_param().is_some() {
+                let this = || make::name_ref("this").syntax().clone_for_update();
+                usages_for_locals(params[0].1.as_local(ctx.sema.db))
+                    .flat_map(|FileReference { name, range, .. }| match name {
+                        ast::NameLike::NameRef(_) => Some(body.syntax().covering_element(range)),
+                        _ => None,
+                    })
+                    .for_each(|it| {
+                        ted::replace(it, &this());
+                    })
+            }
+
+            // Inline parameter expressions or generate `let` statements depending on whether inlining works or not.
+            for ((pat, _), usages, expr) in izip!(params, param_use_nodes, arg_list).rev() {
+                match &*usages {
+                    // inline single use parameters
```

Comment:
this might be a problem if the parameter expression has a side effect?

Reply:
oh dammit right, the evaluation could get skipped now due to early returns. I didn't think of that. Guess this has to be restricted to simple name_ref expressions as well. Though I guess there is also the problem with `drop` impls then?

Looks like we just cannot inline parameters at all then... ?

Reply:
I think inlining simple namerefs should be fine, I don't see how that could lead to problems even with drop? We could even extend it to things like `&name_ref`, probably. (Maybe that could lead to borrow checking problems? Not sure) We could also inline single-use parameters if their expression is the first in the body, but that might be too much of an edge case.

Reply:
I don't think we should try too hard to be semantics-preserving at this stage. 

Long term, it'd be cool to show some kind of UI indicator for cases where we are not sure if semantics is the same.

Reply:
> I don't see how that could lead to problems even with drop? 

It *could* affect drop order I believe.

Though I agree, I don't think people will use this assist on functions where the args they supply will have side effects too often? I feel like you'd usually notice that as you normally dont wanna inline really big functions where you wouldn't notice the inlining happening.

Reply:
> I don't think we should try too hard to be semantics-preserving at this stage.
> 
> Long term, it'd be cool to show some kind of UI indicator for cases where we are not sure if semantics is the same.

I agree that not everything needs to be semantics-preserving, but I think for inlining arguments into the method body we should err on the side of not inlining, because it's easier to use the "inline variable" assist afterwards than to extract and move the variable to the right place.

Reply:
Mm, being able to use `inline variable` is a strong point in favor of not eagerly inlining everything.